### ISSUE TITLE
Async loader and error handling

### ISSUE THREAD

**Issue by mannyvergel** [mozilla/nunjucks#324]
I have an async loader and I passed an error in the callback, however, the behavior I got was unexpected. It just threw the error and it didn't reach my other callback in `environment.render`. I want the error to reach my other callback so I can render a custom response. 

The source code of what I'm doing is in https://github.com/mannyvergel/wateroo-plugin-wcm/blob/master/wcmRoute.js

I checked environment.js starting line 130 and I wanted to change it to following and contribute:

```
        lib.asyncIter(this.loaders, function(loader, i, next, done) {
                function handle(err, src) {
                    if(err || src) {
                        done(err, src);
                    }
                    else {
                        next();
                    }
                }

                if(loader.async) {
                    loader.getSource(name, function(err, src) {
                        handle(err, src);
                    });
                }
                else {
                    handle(null, loader.getSource(name));
                }
            }, function(err2, info) {
                if (err2) {
                  cb(err2);
                }
                else if(!info) {
                    var err = new Error('template not found: ' + name);
                    if(cb) {
                        cb(err);
                    }
                    else {
                        throw err;
                    }
                }
                else {
                    var tmpl = new Template(info.src, this,
                                            info.path, eagerCompile);

                    if(!info.noCache) {
                        this.cache[name] = tmpl;
                    }

                    if(cb) {
                        cb(null, tmpl);
                    }
                    else {
                        syncResult = tmpl;
                    }
                }
            }.bind(this));
```

Basically the change is to pass the error to the callback instead of just throwing it. However, I feel that changing it to the above will make it inconsistent with the "sync" behavior. Any idea, maybe I'm attacking the problem the wrong way.'"
